timemachine/cpp/src/kernels/k_nonbonded_common.cuh

```diff
@@ -12,23 +12,157 @@ static const int NONBONDED_KERNEL_THREADS_PER_BLOCK = 256;
+float __device__ __forceinline__ switch_fn_and_deriv(float dij, float *dsdr) {
+
+    // constants
+    constexpr float cutoff = 1.2;
+    constexpr float pi = static_cast<float>(PI);
+    constexpr float pi_over_2 = 0.5f * pi;
+    constexpr float inv_cutoff = 1 / cutoff;
+
+    // cutoff^-8
+    constexpr float k2 = inv_cutoff * inv_cutoff;
+    constexpr float k4 = k2 * k2;
+    constexpr float k8 = k4 * k4;
+
+    // dij^7, dij^8
+    float dij2 = dij * dij;
+    float dij4 = dij2 * dij2;
+    float dij7 = dij4 * dij2 * dij;
+    float dij8 = dij4 * dij4;
+
+    // arg = (dij/cutoff)^8 * pi/2
+    float dij_k8 = dij8 * k8;
+    float arg = pi_over_2 * dij_k8;
+
+    // sin(arg), cos(arg)
+    float sin_arg;
+    float cos_arg;
+    __sincosf(arg, &sin_arg, &cos_arg);
+
+    // exponentiation
+    float cos_arg2 = cos_arg * cos_arg;
+    float cos_arg3 = cos_arg2 * cos_arg;
+
+    // write d switch_fn d r
+    constexpr float minus_12_pi_k8 = -12 * pi * k8;
+    dsdr[0] = minus_12_pi_k8 * dij7 * sin_arg * cos_arg2;
+
+    // return switch_fn(dij)
+    float sr = cos_arg3;
+    return sr;
+}
+
+float __device__ __forceinline__ fast_erfc_and_deriv(float x, float *dedx) {
+    // TODO: consider using fasterfc implementations listed in this thread:
+    // https://forums.developer.nvidia.com/t/calling-all-juffas-whats-up-with-erfcf-nowadays/262973/4
+
+    float exp_beta_x2 = __expf(-x * x);
+    // (ytz) 5th order gaussian polynomial approximation, we need the exp(-x^2) anyways for the chain rule
     // so we use last variant in https://en.wikipedia.org/wiki/Error_function#Approximation_with_elementary_functions
-    float t = 1.0f / (1.0f + 0.3275911f * beta_dij);
-    erfc_beta_dij = (0.254829592f + (-0.284496736f + (1.421413741f + (-1.453152027f + 1.061405429f * t) * t) * t) * t) *
-                    t * exp_beta_dij_2;
-    return -inv_d2ij * (static_cast<float>(TWO_OVER_SQRT_PI) * beta_dij * exp_beta_dij_2 + erfc_beta_dij);
+    float t = __frcp_rn(1.0f + 0.3275911f * x);
+    float erfc_x = (0.254829592f + (-0.284496736f + (1.421413741f + (-1.453152027f + 1.061405429f * t) * t) * t) * t) *
+                   t * exp_beta_x2;
+    constexpr float minus_two_over_sqrt_pi = -static_cast<float>(TWO_OVER_SQRT_PI);
+    dedx[0] = minus_two_over_sqrt_pi * exp_beta_x2;
+    return erfc_x;
+}
+
+float __device__ __forceinline__
+real_es_factor(float real_beta, float dij, float inv_dij, float inv_d2ij, float &damping_factor) {
+    float beta_dij = real_beta * dij;
+
+    // f(dij) = erfc(beta * dij)
+    // ebd = f(dij), debd = f'(dij)
+    float debd;
+    float ebd = fast_erfc_and_deriv(beta_dij, &debd);
+    debd = real_beta * debd;
+
+    // g(dij) = switch_fn(dij)
+    // sr = g(dij), dsdr = g'(dij)
+    float dsdr;
+    float sr = switch_fn_and_deriv(dij, &dsdr);
+
+    damping_factor = ebd * sr;
+    float damping_factor_prime = (ebd * dsdr) + (debd * sr);
+    float d_es_dr = damping_factor_prime * inv_dij - damping_factor * inv_d2ij;
```

Comment:
this can be written as `inv_dij * (damping_factor_prime  - damping_factor * inv_dij)` which would avoid the need for the outer calling function `compute_electrostatics` to compute and pass in `inv_d2ij` to start with.